packages/modals/examples/tooltip-modal.md

```diff
@@ -0,0 +1,129 @@
+    if (step === 1) {
+      return step1Ref.current;
+    } else if (step === 2) {
+      return step2Ref.current;
+    } else if (step === 3) {
+      return step3Ref.current;
+    } else if (step === 4) {
+      return step4Ref.current;
+    }
+
+    return undefined;
+  }, [step]);
+
+  const placement = React.useMemo(() => {
+    if (step === 1) {
+      return 'bottom-start';
+    } else if (step === 2 || step === 3) {
+      return 'bottom';
+    } else if (step === 4) {
+      return 'bottom-end';
+    }
+
+    return undefined;
+  }, [step]);
+
+  return (
+    <Grid>
+      <Row className="u-mb-md">
+        <Col>
+          <Well isRecessed>
+            <Field>
+              <Toggle checked={hasArrow} onChange={e => setHasArrow(e.target.checked)}>
+                <Label>
+                  <Code>hasArrow</Code>
+                </Label>
+              </Toggle>
+            </Field>
+          </Well>
+        </Col>
+      </Row>
+      <Row>
+        <Col>
+          <Button ref={step1Ref} onClick={() => setStep(1)}>
+            Step 1
+          </Button>
+        </Col>
+        <Col>
+          <Button ref={step2Ref} onClick={() => setStep(2)}>
+            Step 2
+          </Button>
+        </Col>
+        <Col>
+          <Button ref={step3Ref} onClick={() => setStep(3)}>
+            Step 3
+          </Button>
+        </Col>
+        <Col>
+          <Button ref={step4Ref} onClick={() => setStep(4)}>
+            Step 4
+          </Button>
+        </Col>
+      </Row>
+      <TooltipModal
+        referenceElement={referenceElement}
+        placement={placement}
+        hasArrow={hasArrow}
+        onClose={() => setStep(undefined)}
+      >
+        <TooltipModal.Title>Header content</TooltipModal.Title>
+        <TooltipModal.Body>
+          Gumbo beet greens corn soko endive gumbo gourd. Parsley shallot courgette tatsoi pea
+          sprouts fava bean collard greens dandelion okra wakame tomato. Dandelion cucumber earthnut
+          pea peanut soko zucchini.
+        </TooltipModal.Body>
+        <TooltipModal.Footer>
+          <StyledProgress>
+            <span>{step} of 4</span>
+          </StyledProgress>
+          <TooltipModal.FooterItem>
+            <Button size="small" disabled={step === 1} onClick={() => setStep(step - 1)}>
+              Previous
+            </Button>
+          </TooltipModal.FooterItem>
+          <TooltipModal.FooterItem>
+            <Button
+              size="small"
+              isPrimary
+              onClick={() => {
+                if (step === 4) {
+                  setStep(undefined);
+                } else {
+                  setStep(step + 1);
+                }
+              }}
+            >
+              {step === 4 ? 'Finish' : 'Next'}
+            </Button>
+          </TooltipModal.FooterItem>
+        </TooltipModal.Footer>
+        <TooltipModal.Close aria-label="Close" />
+      </TooltipModal>
+    </Grid>
+  );
+};
+
+<Example />;
```

Comment:
This is looking really good! I noticed that when the user presses `esc` in the example, the first trigger always gets focused. Do you think the example can have focus returned to the trigger element which caused the tooltip modal to open?

![2020-08-10 14 30 31](https://user-images.githubusercontent.com/1811365/89833701-57619d80-db16-11ea-9624-c2013a36695c.gif)

Reply:
I’m not sure if this is possible with the logic provided by `useModal`. It’s the same modal throughout all the steps, so the focus is returned to that first triggered element. 

Would it make more sense to set the `restoreFocus` prop as `false` by default?

Reply:
Setting `restoreFocus` to `false` seems like a temporary solution and sounds good to me 👍 .

However, I think setting `restoreFocus` to `false` would not be ideal because the [spec](https://www.w3.org/TR/wai-aria-practices-1.1/#keyboard-interaction-7) says "When a dialog closes, focus returns to the element that invoked the dialog ". But this is an interesting situation because the "element that invoked the dialog" could be located in the most recently closed dialog.

When a user clicks "Step 1" and navigates to the "Step 3" modal tooltip by selecting the "Next" button, then presses `esc` while the "Step 3" modal tooltip is opened, where should the focus be after the modal tooltip is closed?

@m-lai would love to get your opinion on this too!

Reply:
Another wrinkle is that these may be positioned against elements which are not focusable. Like sidebars and the like.

Reply:
Does it make sense to add an optional `restoreFocusRef` prop so that the consumer can determine how restored focus is handled?

Reply:
Possibly, if we are able to disable the default functionality in `useModal`. Most of the "wizard" flows where these are repositioned are programmatically triggered as well. There may not be a sane way to find the correct focus-point without it being provided.

Reply:
Perhaps `useModal` needs to be enhanced to accept a `restoreFocusRef`?

Reply:
I would go against that. For their most common usage, modals are triggered from a single item. This aligns with how `useModal` works now.

The cleanest solution to be feels like we should do the following:
- disable `restoreFocus` in the `useModal()` hook to remove the default functionality
- use `restoreFocus` to apply focus to the current `referenceElement` when the component is unmounted
  - Probably have to do this in a `useEffect()` cleanup

This would still allow consumers to disable the functionality if positioned against a non-focusable element.

Reply:
If you're able to achieve the desired behavior, then this sounds like a preferred approach.

Reply:
The latest commit adds this functionality. Wasn't as bad as I was expecting.

Reply:
Wow, looks great!